cmd/dp-import-tracker/main.go

```diff
@@ -0,0 +1,256 @@
+	for _, instance := range instancesFromAPI {
+		instanceID := instance.InstanceID
+		trackedInstances[instanceID] = trackedInstance{
+			totalObservations:         instance.NumberOfObservations,
+			observationsInsertedCount: instance.TotalInsertedObservations,
+			jobID: instance.Job.ID,
+		}
+	}
+	return nil
+}
+
+// CheckImportJobCompletionState checks all instances for given import job - if all completed, mark import as completed
+func CheckImportJobCompletionState(client *http.Client, importAPIURL string, jobID, completedInstanceID string) error {
+	importJobFromAPI, err := api.GetImportJob(client, importAPIURL, jobID)
+	if err != nil {
+		return err
+	}
+
+	targetState := "completed"
+	log.Debug("checking", log.Data{"API insts": importJobFromAPI.Instances})
+	for _, instance := range importJobFromAPI.Instances {
+		if instance.InstanceID == completedInstanceID {
+			continue
+		}
+		// XXX TODO code below largely untested, and possibly subject to race conditions
+		instanceFromAPI, err := api.GetInstance(client, importAPIURL, instance.InstanceID)
+		if err != nil {
+			return err
+		}
+		if instanceFromAPI.State != "completed" && instanceFromAPI.State != "error" {
+			return nil
+		}
+		if instanceFromAPI.State == "error" {
+			targetState = instanceFromAPI.State
+		}
+	}
+	// assert: all instances for jobID are marked "completed"/"error", so update import as same
+	if err := api.UpdateImportJobState(client, importAPIURL, jobID, targetState); err != nil {
+		log.ErrorC("Failed to set import job state=completed", err, log.Data{"jobID": jobID, "last completed instanceID": completedInstanceID})
+	}
+	return nil
+}
+
+// updateInstanceWithObservationsInserted updates a specific import instance with the counts of inserted observations
+func updateInstanceWithObservationsInserted(client *http.Client, importAPIURL string, instanceID string, observationsInserted int32) error {
+	if err := api.UpdateInstanceWithNewInserts(client, importAPIURL, instanceID, observationsInserted); err != nil {
+		return err
+	}
+	return nil
+}
+
+// manageActiveInstanceEvents handles all updates to trackedInstances in one thread (this is only called once, in its own thread)
+func manageActiveInstanceEvents(createInstanceChan chan string, updateInstanceWithObservationsInsertedChan chan insertedObservationsEvent, importAPIURL string, client *http.Client) {
+	trackedInstances := make(trackedInstanceList)
+	if err := trackedInstances.getInstanceListFromImportAPI(client, importAPIURL); err != nil {
+		logFatal("Could not obtain initial instance list", err, nil)
+	}
+
+	checkForCompletedInstancesChan := make(chan bool)
+	go func() {
+		for range time.Tick(checkForCompleteInstancesTick) {
+			checkForCompletedInstancesChan <- true
+		}
+	}()
+
+	for {
+		select {
+		case newInstanceMsg := <-createInstanceChan:
+			if _, ok := trackedInstances[newInstanceMsg]; ok {
+				log.Error(errors.New("import instance exists"), log.Data{"instanceID": newInstanceMsg})
```

Comment:
Generally a break/continue might be clearer than the else - keeps the happy path clearer